[language] Add changelog for phantom types (#8827)
diem/diem: language/changes/6-phantom-type-params.md

# Phantom Type Parameters

- Status: Implemented in Move 1.4

## Introduction

A phantom type parameter is one that doesn't show up at runtime, but is checked statically at
compile time. Previously, Move's type system didn't make a difference on whether a type parameter
was phantom, but with this new feature, a struct's type parameter can be explicitly declared as
phantom. A parameter declared as phantom is not considered when computing the
[conditional abilities with generic types](3-abilities.md#Conditional-Abilities-with-Generic-Types).
For this relaxed rule to be sound, Move's type system guarantees that a parameter declared as
phantom is either not used at all in the struct definition, or it is only used as an argument to
type parameters also declared as phantom.

## Motivations

> A phantom type parameter may come in handy in a couple of situations. For example, the struct

Reviewer: This first paragraph in the Motivations section seems misplaced. It says nothing about the motivation for phantom type parameters, but instead offers examples from the Diem Framework where they can be used. We should move this elsewhere. Since the other Motivations paragraph refers to these specific examples, I suggest moving this paragraph into the Introduction section.

A few other things:

- Let's also drop the first sentence. This is more substantial than providing something "handy".
- Replace "stdlib" with "the Move standard library"
- The term "notorious" livens up the tone of this document but seems too strong. Maybe "prominent"?

`Event::EventHandle<T>` in stdlib doesn't contain any field of type `T`, but uses the type parameter
to ensure at the type level that the handle can only be used for messages of type `T`. Another
notorious example is `Diem::Diem<CoinType>`, which is generic on a `CoinType` specifying the
currency of the coin and allowing code to be written generically on any currency. As a last example,
the capabilities in `Vault` also rely heavily on phantom type parameters.

> Previously, these examples required spurious ability annotations to satisfy the requirements of the

Reviewer: ```suggestion
Previously, phantom type parameters required spurious ability annotations to satisfy the requirements of the
```

abilities declared for each of the structs. For example, `Diem::Diem<CoinType>` is declared with the
ability `key`, which required all arguments to `CoinType` to be declared with `store`. Specifically,
the types `XUS::XUS` and `XDX::XDX` were forced to have a `store` annotation even though they were
never stored in the database! More over the extra `store` requirement polluted the entire Diem
Framework and many functions had to specify an extra `store` constraints on their type parameters.

> With the new feature, `Diem::Diem<CoinType>` now declares `CoinType` to be phantom which avoids the

Reviewer: It would be better to first state this generally, i.e., saying that this language change avoids the need for spurious abilities, before giving the specific example of `Diem<CoinType>`. It might also be good to add something before this concluding sentence to elaborate on why the spurious abilities are bad.

unnecessary `store` annotations.

## Description

### Declaration

In a struct definition a type parameter can be declared as phantom by adding the `phantom` keyword
before its declaration. If a type parameter is declared as phantom we say it is a phantom type
parameter. When defining a struct, Move's type checker ensures that every phantom type parameter is
either not used inside the struct definition or it is only used as an argument to a phantom type
parameter.

More formally, if a type is used as an argument to a parameter declared as phantom we say the type
appears in phantom position. With this definition in place, the rule for the correct use of phantom
parameters can be specified as follows: **A phantom type parameter can only appear in phantom
position**.

The following two examples show valid uses of phantom parameters. In the first one, the parameter
`T1` is not used at all inside the struct definition. In the second one, the parameter `T1` is only

> used as an argument to a parameter also declared as phantom.

Reviewer: I find this "declared as phantom" terminology awkward. I don't have a good suggestion for a direct replacement, but we could try to avoid using it more than necessary. For example, this sentence could say "...used as an argument to another phantom type parameter." Perhaps you could rewrite some of the other text with this terminology in a similar fashion?

struct S2<phantom T1, T2> { f: S1<T1, T2> }
                                  ^^
                                  Ok: T1 appears in phantom position
```

The following code shows examples of violations of the rule:

```

struct S1<phantom T> { f: T }
                          ^
                          Error: Not a phantom position

struct S2<T> { f: T }

struct S3<phantom T> { f: S2<T> }
                             ^
                             Error: Not a phantom position
```

### Instantiation

When instantiating a struct, the arguments to phantom parameters are excluded when deriving the
struct abilities. For example, consider the following code:

```
struct S<T1, phantom T2> has copy { f: T1 }
struct NoCopy {}
struct HasCopy has copy {}
```

Consider now the type `S<HasCopy, NoCopy>`. Since `S` is defined with `copy` and all non-phantom
arguments have copy then `S<HasCopy, NoCopy>` also has copy.

### Phantom Type Parameters with Ability Constraints

Ability constraints and phantom type parameters are orthogonal features in the sense that phantom
parameters can be declared with ability constraints. And any type parameter must satisfy that

> constraint, even thought it is phantom. For example, the following definition is perfectly valid:

Reviewer: A few nits on this sentence:
- typo: "thought" should be "though
- It's not a complete sentence. At a minimum, you need a comma after the initial "and", but probably there's a better way.
- It is not clear what "that constraint" refers to. The preceding sentence refers to "ability constraints" in general but says nothing about a specific constraint.
- It is also not clear what "it" refers to. Presumably "any type parameter", but it's confusing because it looks like it could be "that constraint" (which makes no sense).

struct S2<phantom T1, T2> { f: S1<T1, T2> }
                                  ^^
                                  Ok: T1 appears in phantom position
```

The following code shows examples of violations of the rule:

```

struct S1<phantom T> { f: T }
                          ^
                          Error: Not a phantom position

struct S2<T> { f: T }

struct S3<phantom T> { f: S2<T> }
                             ^
                             Error: Not a phantom position
```

### Instantiation

When instantiating a struct, the arguments to phantom parameters are excluded when deriving the
struct abilities. For example, consider the following code:

```
struct S<T1, phantom T2> has copy { f: T1 }
struct NoCopy {}
struct HasCopy has copy {}
```

Consider now the type `S<HasCopy, NoCopy>`. Since `S` is defined with `copy` and all non-phantom
arguments have copy then `S<HasCopy, NoCopy>` also has copy.

### Phantom Type Parameters with Ability Constraints

Ability constraints and phantom type parameters are orthogonal features in the sense that phantom
parameters can be declared with ability constraints. When instantiating a phantom type parameter

> with an ability constraints, the type argument has to satisfy that constraint, even though the

Reviewer: ```suggestion
with an ability constraint, the type argument has to satisfy that constraint, even though the
```

but uses the type parameter to ensure at the type level that the handle can only be used for
messages of type `T`. Another prominent example is `Diem::Diem<CoinType>`, which is generic on a
`CoinType` specifying the currency of the coin and allowing code to be written generically on any
currency. As a last example, the capabilities in `Vault` also rely heavily on phantom type
parameters.

Previously, Move's type system didn't make a difference on whether a type parameter was phantom, but
with this new feature, a struct's type parameter can be explicitly declared as phantom. A parameter
declared as phantom is not considered when computing the
[conditional abilities with generic types](3-abilities.md#Conditional-Abilities-with-Generic-Types).
For this relaxed rule to be sound, Move's type system guarantees that a parameter declared as
phantom is either not used at all in the struct definition, or it is only used as an argument to
type parameters also declared as phantom.

## Motivations

Previously, defining a struct `S` with a phantom type parameter required spurious ability
annotations to satisfy the requirements of the abilities declared for `S`. For a module author, this

> meant loosing control over structs that had to be declared with unnecessary abilities. Moreover, the

Reviewer: ```suggestion
meant losing control over structs that had to be declared with unnecessary abilities. Moreover, the
```

I'm not exactly sure what "losing control" means here. I was expecting something more along the lines of increased chance for bugs or security holes by weakening the type checking because of the unnecessary abilities.
